[CLIPPER-129] Update RPC benchmark to use event-based RPC service

dcrankshaw opened the pull request:
This also cleans up the benchmark to make it a little cleaner and easier to use.

dcrankshaw committed:
got new benchmark compiling

dcrankshaw committed:
Revised RPC benchmark is working

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/74/
Test PASSed.

dcrankshaw committed:
more informative error message in rpc bench

dcrankshaw committed:
sum responses in rpc benchmark

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/75/
Test PASSed.

Corey-Zumar commented on src/benchmarks/src/rpc_service_bench.cpp:
```diff
@@ -0,0 +1,286 @@
+        "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+        "abcdefghijklmnopqrstuvwxyz";
+    const size_t max_index = (sizeof(charset) - 1);
+    return charset[rand() % max_index];
```
We need to seed the generator before `rand()` is called. `srand(time(NULL));`, for example

dcrankshaw replied:
Good catch

Corey-Zumar commented on src/benchmarks/src/rpc_service_bench.cpp:
```diff
@@ -0,0 +1,286 @@
+}
+
+rpc::PredictionRequest create_request(InputType input_type) {
+  int num_inputs = 500;
```
We should make `num_inputs` a parameter.

dcrankshaw replied:
Done. Make it a command line parameter.

Corey-Zumar commented on src/benchmarks/src/rpc_service_bench.cpp:
```diff
@@ -0,0 +1,286 @@
+                          InputType::Doubles);
+  benchmarker.start();
+
+  // auto jh = std::thread([&benchmarker]() { benchmarker.start(); });
```
Remove commented code

Corey-Zumar commented on src/benchmarks/src/rpc_service_bench.cpp:
```diff
@@ -0,0 +1,286 @@
+  std::unique_lock<std::mutex> l(benchmarker.bench_completed_cv_mutex);
+  benchmarker.bench_completed_cv_.wait(
+      l, [&benchmarker]() { return benchmarker.bench_completed_ == true; });
+  // jh.join();
```
Remove commented code

dcrankshaw replied:
Done

Corey-Zumar commented on src/benchmarks/src/rpc_service_bench.cpp:
```diff
@@ -0,0 +1,286 @@
+  }
+
+  std::condition_variable_any bench_completed_cv_;
+  std::mutex bench_completed_cv_mutex;
```
This should have a trailing underscore: `bench_completed_cv_mutex_`

dcrankshaw replied:
Good catch.

Corey-Zumar commented on bench/README.md:
````diff
@@ -37,24 +37,24 @@ If a configuration file is not specified, the benchmark will prompt you for the
 1. Execute the following:
-  ```
-  $ ./setup_bench.sh <path_to_cifar_python_dataset>
+  ```sh
````
That's really cool. Didn't know about `sh`

dcrankshaw replied:
You can also use ` ```console ` or ` ```bash `

Corey-Zumar commented on src/benchmarks/src/rpc_service_bench.cpp:
```diff
@@ -0,0 +1,286 @@
+  conf.set_redis_port(options["redis_port"].as<int>());
+  conf.ready();
+  Benchmarker benchmarker(options["num_messages"].as<int>(),
+                          InputType::Doubles);
```
We should add an argument allowing us to specify the input type.

dcrankshaw replied:
Done

Corey-Zumar commented on src/benchmarks/src/rpc_service_bench.cpp:
```diff
@@ -0,0 +1,286 @@
+    }
+    log_info_formatted(LOGGING_TAG_RPC_BENCH, "output sum: {}", sum);
+    msg_latency_hist_->insert(message_duration_millis);
+    throughput_meter_->mark(1);
```
Because we only send a new message when a response is received, it seems that our throughput will always be relatively low.

dcrankshaw replied:
So that's actually how the containers work. Each container can only handle one message at a time. So this should reflect the same throughput as the full system. Keep in mind that we're sending mini-batches though, so each message contains many requests.

Corey-Zumar commented on src/benchmarks/src/rpc_service_bench.cpp:
```diff
@@ -0,0 +1,286 @@
+  }
+
+  void on_response_recv(rpc::RPCResponse response) {
+    // process the response
```
Remove comment

dcrankshaw replied:
Done.

Corey-Zumar commented on src/benchmarks/src/rpc_service_bench.cpp:
```diff
@@ -0,0 +1,286 @@
+      sum += d;
+    }
+  }
+  log_info_formatted(LOGGING_TAG_RPC_BENCH, "Request inputs sum: {}", sum);
```
Why are we computing and logging this sum for doubles?

dcrankshaw replied:
I was using it as a checksum to make sure that the Java container implementation was working correctly. I'll remove it.

Corey-Zumar reviewed (changes requested):
Looks good overall, mostly nits and doc requests.

dcrankshaw committed:
Addressed review comments and ran the code formatter which cleaned up some whitespace issues

dcrankshaw reviewed (commented):
Addressed review comments

dcrankshaw commented:
@Corey-Zumar addressed your comments and tested it with the Python Noop container.

Sorry for polluting the PR a bit with whitespace changes, I guess I still haven't fully worked out the format checker in the CI script.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/80/
Test PASSed.

Corey-Zumar reviewed (approved):
LGTM

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/84/
Test FAILed.

dcrankshaw committed:
fixed format check errors

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/85/
Test PASSed.

dcrankshaw commented:
@Corey-Zumar Fixed formatting so the checks all pass. Merge this when you get a chance.
